src/track/beatmap.cpp

```diff
@@ -315,80 +315,53 @@ audio::FramePos BeatMap::findNthBeat(audio::FramePos position, int n) const {
-    BeatList::const_iterator next_beat = m_beats.constEnd();
-    for (; it != m_beats.end(); ++it) {
-        qint32 delta = it->frame_position() - beat.frame_position();
+    if (n > 0) {
+        // Search in forward direction
+        numBeatsLeft--;
 
-        // We are "on" this beat.
-        if (abs(delta) < kFrameEpsilon) {
-            on_beat = it;
-            break;
+        while (it != m_beats.cend() && numBeatsLeft > 0) {
+            if (it->enabled()) {
+                numBeatsLeft--;
+            }
+            it++;
+        }
+    } else {
+        // Search in backward direction
+        numBeatsLeft--;
+
+        if (it == m_beats.cend() || it->frame_position() > searchFromBeat.frame_position()) {
+            // We may be one beat behind the searchFromBeat. In this case, we advance
+            // the reverse iterator by one beat.
+            if (it == m_beats.cbegin()) {
+                return audio::kInvalidFramePos;
+            }
+            it--;
         }
 
-        if (delta < 0) {
-            // If we are not on the beat and delta < 0 then this beat comes
-            // before our current position.
-            previous_beat = it;
-        } else {
-            // If we are past the beat and we aren't on it then this beat comes
-            // after our current position.
-            next_beat = it;
-            // Stop because we have everything we need now.
-            break;
+        while (it != m_beats.cbegin() && numBeatsLeft > 0) {
+            if (it->enabled()) {
+                numBeatsLeft--;
+            }
+            it--;
         }
     }
 
-    // If we are within epsilon samples of a beat then the immediately next and
-    // previous beats are the beat we are on.
-    if (on_beat != m_beats.end()) {
-        next_beat = on_beat;
-        previous_beat = on_beat;
+    if (numBeatsLeft > 0 || it == m_beats.cend()) {
+        return audio::kInvalidFramePos;
     }
 
-    if (n > 0) {
-        for (; next_beat != m_beats.end(); ++next_beat) {
-            if (!next_beat->enabled()) {
-                continue;
-            }
-            if (n == 1) {
-                return mixxx::audio::FramePos(next_beat->frame_position());
-            }
-            --n;
-        }
-    } else if (n < 0 && previous_beat != m_beats.end()) {
-        for (; true; --previous_beat) {
-            if (previous_beat->enabled()) {
-                if (n == -1) {
-                    return mixxx::audio::FramePos(previous_beat->frame_position());
-                }
-                ++n;
-            }
+    const auto foundBeatPosition = mixxx::audio::FramePos(it->frame_position());
 
-            // Don't step before the start of the list.
-            if (previous_beat == m_beats.begin()) {
-                break;
-            }
-        }
-    }
-    return audio::kInvalidFramePos;
+    DEBUG_ASSERT(foundBeatPosition >= searchFromPosition || n < 0);
```

Comment:
there's a lot of "n > 0" repeated in this function, can we factor that out to "const bool searchForward = n > 0;"?

Reply:
Good idea.

Reply:
9037bb55e68f5e582510cc6cefac066f917c483c